File: algorithms/167.Two_Sum_II_-_Input_array_is_sorted/Two_Sum_II_-_Input_array_is_sorted.py
# Time: O(n)
# Space: O(1)
class Solution:
    # @param {integer[]} numbers
    # @param {integer} target
    # @return {integer[]}
    def twoSum(self, numbers, target):
        start, end = 0, len(numbers) - 1

        while start != end:
            sum = numbers[start] + numbers[end]
            if sum > target:
                end -= 1
            elif sum < target:
                start += 1
            else:
                return [start + 1, end + 1]

if __name__ == "__main__":
    print Solution().twoSum([2, 7, 11, 15], 9)

File: algorithms/167.Two_Sum_II_-_Input_array_is_sorted/Two_Sum_II_-_Input_array_is_sorted.cpp
// https://leetcode.com/problems/two-sum-ii-input-array-is-sorted/
// O(n) time, O(1) space
class Solution {
 public:
  vector<int> twoSum(vector<int>& numbers, int target) {
    vector<int> result;
    int s = 0, e = numbers.size() - 1;
    while(s < e) {
      int v = numbers[s] + numbers[e];
      if(v == target) {
        result.push_back(s+1);
        result.push_back(e+1);
        return result;
      }
      else if(v < target) {
        s += 1;
      }
      else {
        e -= 1;
      }
    }
  }
};